From 5dfff828bc1afbe66ebc6c4e890b4a937ab4a934 Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Fri, 5 May 2023 10:29:57 +0800
Subject: [PATCH 1602/1646] [Huawei] userswap: split uswap_register() to
 validate address ranges

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6CAIM

--------------------------------

Split uswap_register() into uswap_register() and uswap_adjust_uffd_range().
Before validate_range(), use uswap_register() to handle uswap mode.
After validate_range(), use uswap_adjust_uffd_range() to change address
range to VMA range, which could reduce fragmentation caused by VMA
splitting.
By splitting uswap_register(), we could prevent the userswap registration
of invalid input address ranges.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 fs/userfaultfd.c         | 14 ++++++++++++--
 include/linux/userswap.h |  6 ++++--
 mm/userswap.c            | 26 +++++++++++++++++---------
 3 files changed, 33 insertions(+), 13 deletions(-)

diff --git a/fs/userfaultfd.c b/fs/userfaultfd.c
index 1222e507fade..ee20b9620ac2 100644
--- a/fs/userfaultfd.c
+++ b/fs/userfaultfd.c
@@ -1282,6 +1282,9 @@ static int userfaultfd_register(struct userfaultfd_ctx *ctx,
 	bool found;
 	bool basic_ioctls;
 	unsigned long start, end, vma_end;
+#ifdef CONFIG_USERSWAP
+	bool uswap_mode = false;
+#endif
 
 	user_uffdio_register = (struct uffdio_register __user *) arg;
 
@@ -1295,7 +1298,7 @@ static int userfaultfd_register(struct userfaultfd_ctx *ctx,
 		goto out;
 	vm_flags = 0;
 #ifdef CONFIG_USERSWAP
-	if (!uswap_register(&uffdio_register, &vm_flags, mm))
+	if (!uswap_register(&uffdio_register, &uswap_mode))
 		goto out;
 #endif
 	if (uffdio_register.mode & ~(UFFDIO_REGISTER_MODE_MISSING|
@@ -1310,7 +1313,14 @@ static int userfaultfd_register(struct userfaultfd_ctx *ctx,
 			     uffdio_register.range.len);
 	if (ret)
 		goto out;
-
+#ifdef CONFIG_USERSWAP
+	if (uswap_mode)
+		if (!uswap_adjust_uffd_range(&uffdio_register, &vm_flags,
+					     mm)) {
+			ret = -EINVAL;
+			goto out;
+		}
+#endif
 	start = uffdio_register.range.start;
 	end = start + uffdio_register.range.len;
 
diff --git a/include/linux/userswap.h b/include/linux/userswap.h
index 94d773364618..611f84bf3569 100644
--- a/include/linux/userswap.h
+++ b/include/linux/userswap.h
@@ -28,8 +28,10 @@ int mfill_atomic_pte_nocopy(struct mm_struct *dst_mm,
 unsigned long uswap_mremap(unsigned long old_addr, unsigned long old_len,
 			   unsigned long new_addr, unsigned long new_len);
 
-bool uswap_register(struct uffdio_register *uffdio_register,
-		    unsigned long *vm_flags, struct mm_struct *mm);
+bool uswap_register(struct uffdio_register *uffdio_register, bool *uswap_mode);
+
+bool uswap_adjust_uffd_range(struct uffdio_register *uffdio_register,
+			     unsigned long *vm_flags, struct mm_struct *mm);
 
 bool do_uswap_page(swp_entry_t entry, struct vm_fault *vmf,
 		   struct vm_area_struct *vma, vm_fault_t *ret);
diff --git a/mm/userswap.c b/mm/userswap.c
index 9e2d063952ff..ff98190eb258 100644
--- a/mm/userswap.c
+++ b/mm/userswap.c
@@ -11,6 +11,7 @@
 #include <linux/mmu_notifier.h>
 #include <linux/userswap.h>
 #include <linux/userfaultfd_k.h>
+#include <linux/security.h>
 
 #include "internal.h"
 
@@ -462,16 +463,8 @@ int mfill_atomic_pte_nocopy(struct mm_struct *mm,
 	return ret;
 }
 
-/*
- * register the whole vma overlapping with the address range to avoid splitting
- * the vma.
- */
-bool uswap_register(struct uffdio_register *uffdio_register,
-		    unsigned long *vm_flags, struct mm_struct *mm)
+bool uswap_register(struct uffdio_register *uffdio_register, bool *uswap_mode)
 {
-	struct vm_area_struct *vma;
-	unsigned long end;
-
 	if (!static_branch_unlikely(&userswap_enabled))
 		return true;
 	if (!(uffdio_register->mode & UFFDIO_REGISTER_MODE_USWAP))
@@ -479,7 +472,22 @@ bool uswap_register(struct uffdio_register *uffdio_register,
 	uffdio_register->mode &= ~UFFDIO_REGISTER_MODE_USWAP;
 	if (!uffdio_register->mode)
 		return false;
+	*uswap_mode = true;
+	return true;
+}
 
+/*
+ * register the whole vma overlapping with the address range to avoid splitting
+ * the vma which could reduce fragmentation.
+ */
+bool uswap_adjust_uffd_range(struct uffdio_register *uffdio_register,
+			     unsigned long *vm_flags, struct mm_struct *mm)
+{
+	struct vm_area_struct *vma;
+	unsigned long end;
+
+	if (!static_branch_unlikely(&userswap_enabled))
+		return true;
 	end = uffdio_register->range.start + uffdio_register->range.len - 1;
 	vma = find_vma(mm, uffdio_register->range.start);
 	if (!vma)
-- 
2.17.1

